core/lib/zksync_core/src/native_erc20_fetcher/mod.rs

```diff
@@ -0,0 +1,115 @@
+use std::{borrow::BorrowMut, collections::HashMap, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use metrics::atomics::AtomicU64;
+use tokio::{
+    sync::{watch, Mutex, OnceCell},
+    task::JoinHandle,
+};
+use zksync_config::configs::native_erc20_fetcher::NativeErc20FetcherConfig;
+
+// TODO: this error type is also defined by the gasAdjuster module,
+//       we should consider moving it to a common place
+#[derive(thiserror::Error, Debug, Clone)]
+#[error(transparent)]
+pub struct Error(Arc<anyhow::Error>);
+
+impl From<anyhow::Error> for Error {
+    fn from(err: anyhow::Error) -> Self {
+        Self(Arc::new(err))
+    }
+}
+
+#[async_trait]
+pub trait Erc20Fetcher: 'static + std::fmt::Debug + Send + Sync {
+    async fn conversion_rate(&self) -> anyhow::Result<u64>;
+}
+
+pub(crate) struct NativeErc20FetcherSingleton {
+    native_erc20_fetcher_config: NativeErc20FetcherConfig,
+    singleton: OnceCell<Result<Arc<NativeErc20Fetcher>, Error>>,
+}
+
+impl NativeErc20FetcherSingleton {
+    pub fn new(native_erc20_fetcher_config: NativeErc20FetcherConfig) -> Self {
+        Self {
+            native_erc20_fetcher_config,
+            singleton: OnceCell::new(),
+        }
+    }
+
+    pub async fn get_or_init(&mut self) -> Result<Arc<NativeErc20Fetcher>, Error> {
+        let adjuster = self
+            .singleton
+            .get_or_init(|| async {
+                let fetcher = NativeErc20Fetcher::new(self.native_erc20_fetcher_config.clone());
+                Ok(Arc::new(fetcher))
+            })
+            .await;
+        adjuster.clone()
+    }
+
+    pub fn run_if_initialized(
+        self,
+        stop_signal: watch::Receiver<bool>,
+    ) -> Option<JoinHandle<anyhow::Result<()>>> {
+        let fetcher = self.singleton.get()?.clone();
+        Some(tokio::spawn(async move { fetcher?.run(stop_signal).await }))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NativeErc20Fetcher {
+    // TODO: we probably need to add a http client here
+    // to avoid creating a new one for each request
+    pub config: NativeErc20FetcherConfig,
+    pub latest_to_eth_conversion_rate: AtomicU64,
+}
+
+impl NativeErc20Fetcher {
+    pub(crate) fn new(config: NativeErc20FetcherConfig) -> Self {
+        Self {
+            config,
+            latest_to_eth_conversion_rate: AtomicU64::new(0),
+        }
+    }
+
+    pub(crate) async fn run(&self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, eth_tx_manager is shutting down");
+                break;
+            }
+
+            let conversion_rate = reqwest::get(format!("{}/conversion_rate", &self.config.host))
```

Comment:
Thanks for the PR @jrchatruc 

Do you think we can put the whole url `{}/conversion_rate` into the configuration instead of hardcoding it in the code?

Reply:
Make use of an interface with the method 
```
getConversionRate() -> u64
```

and have a default implementation that can be overrided by hyper chain owner would help to move toward modularisation of the zkstack

Reply:
Hi @thomas-nguy, when discussing the design with Matter Labs we considered this option, but ultimately decided to go with this approach, where we assume a certain interface/api response format from users. If theirs is slightly different from what we expect, they can simply make a small app/binary that wraps around to conform to our interface.

Tagging @mm-zk for visibility

Reply:
We would like to keep the conversion rate API as a separate binary so that users who want to spin up a stack with a custom base token can provide their conversion rate api in the programming language they prefer, deploying it as they wish. Having it be another component ties you to Rust and the operator deployment.

We can add a simple local development feature where if users do not specify any host in the `erc20_fetcher.toml` file, a testing/mocked fetcher is created that just returns a hardcoded conversion rate.

As for the API design, I agree that it's better to have it be `/conversion_rate/{base_token_addr}/`, we can change it. My only question would be about this:

> e.g. I think it's easy to make the server support multiple base tokens if required

What would supporting multiple base tokens mean? How would that work?

Reply:
> What would supporting multiple base tokens mean? How would that work?

It would mean that you can write a generic binary (e.g. a wrapper for Coingecko api) that supports many tokens, and just deploy it once. You will also be able to ship it as a standalone product that you can use as a configless docker image.

Look, for example, at a [CoingGecko wrapper from zkSync Lite](https://github.com/matter-labs/zksync/blob/master/core/bin/zksync_api/src/fee_ticker/ticker_api/coingecko.rs) -- it can fetch info on any token.

E.g. any particular hyperchain would be interested in just one token, but N hyperchains would be interested in N tokens that can be served by a single binary (instead of either having N binaries or having N configs).

Reply:
> Having it be another component ties you to Rust and the operator deployment.

While true, it doesn't have to be a bad thing -- e.g. for localhost development, you may probably prefer to have a single binary ([having to spawn ](https://github.com/matter-labs/zksync/blob/master/docker-compose.yml#L33) a dev token info server in zkSync Lite wasn't the smoothest local experience). It has both pros and cons.

Note however that I didn't suggest reworking the _fetcher_ as a Rust-only component, I've mentioned that we can introduce a _dev_ component that would only run on localhost and serve some static info on tokens -- this way we will avoid to spawn an extra image. Anyone who is not interested in such a component, will not run it. Similarly, it won't be run in production environment -- some "real" server would run there.

However that's probably out of scope for this PR, as it focuses on the fetcher, not on the server, so we can discuss it further later.

Reply:
Got it, sounds reasonable. Tracked the API endpoint change [here](https://github.com/lambdaclass/zksync-era/issues/198) and the new __dev__ component [here](https://github.com/lambdaclass/zksync-era/issues/199) (even if we don't tackle it on this PR we'll track it)